xtask/src/ci/target_matrix.rs

```diff
@@ -0,0 +1,152 @@
+            platforms: target.platforms(),
+            target: &target.target,
+            sub: target.sub.as_deref(),
+            os: &target.os,
+            run: target.run.map(|b| b as u8),
+            deploy: target.deploy.map(|b| b as u8),
+            build_std: target.build_std.map(|b| b as u8),
+            cpp: target.cpp.map(|b| b as u8),
+            dylib: target.dylib.map(|b| b as u8),
+            runners: target.runners.as_deref(),
+            std: target.std.map(|b| b as u8),
+        })
+        .collect::<Vec<_>>();
+    gha_output("matrix", &serde_json::to_string(&matrix)?);
+    Ok(())
+}
+
+#[derive(Serialize)]
+#[serde(deny_unknown_fields, rename_all = "kebab-case")]
+struct Matrix<'a> {
+    pretty: String,
+    platforms: &'a [String],
+    target: &'a str,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    sub: Option<&'a str>,
+    os: &'a str,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    run: Option<u8>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    deploy: Option<u8>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    build_std: Option<u8>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    cpp: Option<u8>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    dylib: Option<u8>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    runners: Option<&'a str>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    std: Option<u8>,
+}
+
+#[derive(Parser, Debug)]
+#[clap(no_binary_name = true)]
+struct TargetMatrixArgs {
+    #[clap(long, short, multiple_values = true)]
+    target: Vec<String>,
+    #[clap(long, value_parser = BoolishValueParser::new())]
+    std: Option<bool>,
+    #[clap(long, value_parser = BoolishValueParser::new())]
+    cpp: Option<bool>,
+    #[clap(long, value_parser = BoolishValueParser::new())]
+    dylib: Option<bool>,
+    #[clap(long, value_parser = BoolishValueParser::new())]
+    run: Option<bool>,
+    #[clap(long, short, multiple_values = true)]
+    runners: Vec<String>,
+}
+
+impl TargetMatrixArgs {
+    pub fn filter(&self, matrix: &mut Vec<CiTarget>) {
+        if !self.target.is_empty() {
+            matrix.retain(|m| {
+                let matrix_target = m.to_image_target();
+                let matrix_string = matrix_target.to_string();
+
+                return self
+                    .target
+                    .iter()
+                    .any(|t| t.parse::<ImageTarget>().unwrap() == matrix_target)
+                    || self
+                        .target
+                        .iter()
+                        .any(|t| wildmatch::WildMatch::new(t).matches(&matrix_string));
+            })
+        };
+        if let Some(std) = self.std {
+            matrix.retain(|m| m.std.unwrap_or_default() == std)
+        }
+        if let Some(cpp) = self.cpp {
+            matrix.retain(|m| m.cpp.unwrap_or_default() == cpp)
+        }
+        if let Some(dylib) = self.dylib {
```

Comment:
We should probably add a single test to ensure `dylib` or at least one of these checks works, by checking an item in `dylib` (`aarch64-unknown-linux-gnu`) and an item not in `dylib` (`thumbv6m-none-eabi`) are in and not in it respectively. Not all of these code paths need to be tested, but I think at least one should be.